File: draft_design_pattern/simple_factory.nim
type
  Door = ref object of RootObj
  WoodenDoor = ref object of Door
    width: float
    height: float

  DoorFactory = ref object


method getWidth(self: Door): float{.base, locks: "unknown".} =
  discard

method getHeight(self: Door): float{.base, locks: "unknown".} =
  discard


method getWidth(self: WoodenDoor): float{.locks: "unknown".} =
  return self.width

method getHeight(self: WoodenDoor): float{.locks: "unknown".} =
  return self.height


proc makeDoor(self: DoorFactory, width: float, height: float): Door =
  return WoodenDoor(width: width, height: height)


proc main() =
  let obj = DoorFactory()
  let door: Door = obj.makeDoor(200, 300)
  echo door.getWidth()
  echo door.getHeight()

main()

File: draft_design_pattern/simple_factory.py
# シンプルなファクトリは、
# インスタンス化ロジックをクライアントに公開せずに、
# クライアントのインスタンスを生成するだけです。
from abc import ABCMeta, abstractmethod


class Door(metaclass=ABCMeta):
    @abstractmethod
    def getWidth(self) -> float:
        pass

    @abstractmethod
    def getHeight(self) -> float:
        pass


class WoodenDoor(Door):
    def __init__(self, width: float, height: float) -> None:
        self.width: float = width
        self.height: float = height

    def getWidth(self) -> float:
        return self.width

    def getHeight(self) -> float:
        return self.height


class DoorFactory:
    @staticmethod
    def makeDoor(width: float, height: float) -> Door:
        return WoodenDoor(width, height)


door = DoorFactory.makeDoor(100, 200)
print(door.getWidth())
print(door.getHeight())
